mbedtls/src/cipher/raw/serde.rs

```diff
@@ -167,36 +189,7 @@ impl<'de, Op: Operation> Deserialize<'de> for Cipher<Op, Traditional, CipherData
 
         unsafe {
             let cipher_context = &mut raw_cipher.inner;
-
-            match (raw.cipher_id, raw.algorithm_ctx) {
-                (CIPHER_ID_AES, AlgorithmContext::Aes(Bytes(aes_ctx))) => {
-                    let ret_aes_ctx = cipher_context.cipher_ctx as *mut aes_context;
-                    *ret_aes_ctx = aes_ctx;
-                    // aes_ctx.rk needs to be a pointer to aes_ctx.buf, which holds the round keys.
-                    // We don't adjust for the padding needed on VIA Padlock (see definition of
-                    // mbedtls_aes_context in the mbedTLS source).
-                    (*ret_aes_ctx).rk = &mut (*ret_aes_ctx).buf[0];
-                }
-                (CIPHER_ID_DES, AlgorithmContext::Des(Bytes(des_ctx))) => {
-                    *(cipher_context.cipher_ctx as *mut des_context) = des_ctx
-                }
-                (CIPHER_ID_3DES, AlgorithmContext::Des3(Bytes(des3_ctx))) => {
-                    *(cipher_context.cipher_ctx as *mut des3_context) = des3_ctx
-                }
-                _ => {
-                    return Err(de::Error::invalid_value(
-                        Unexpected::Other("bad algorithm"),
-                        &"valid algorithm",
-                    ));
-                }
-            };
-
-            cipher_context.key_bitlen = raw.context.0.key_bitlen;
-            cipher_context.operation = raw.context.0.operation;
-            cipher_context.unprocessed_data = raw.context.0.unprocessed_data;
-            cipher_context.unprocessed_len = raw.context.0.unprocessed_len;
-            cipher_context.iv = raw.context.0.iv;
-            cipher_context.iv_size = raw.context.0.iv_size;
+            deserialize_raw_cipher::<D>(raw, cipher_context)?;
```

Comment:
Can you move lines 167-188 into deserialize_raw_cipher as well so you don't have to repeat cipher setup for GCM internals?